formsfeeder.client/src/main/java/formsfeeder/client/support/Builder.java

```diff
@@ -21,10 +22,16 @@
 
 	public Builder useSsl(boolean useSsl);
 
+	public Builder contextRoot(String contextRoot);
+
 	public Builder clientFactory(Supplier<Client> clientFactory);
 
 	public Builder basicAuthentication(String username, String password);
 
+	public Builder apikeyAuthentication(String apikeyHeader, String apikeyValue);
+
+	public Supplier<Map.Entry<String,String>> getApikeyAuthenticationFn();
+

```

Comment:
### Setter/Getter Asymmetry
What immediately jumps out to me is the asymmetry here – using a name/value pair for the setter but getting a Supplier for the getter.  Is there a reason for this?

The reason I use a `Supplier<>` for correlationId is because I am expecting that the client we’re building will be built once and potentially used multiple times.  Each different invocation might have a different correlation id, so we defer the process of locking in the correlation id until invocation time (rather than at client creation) by using a `Supplier`.

I’d prefer to see the getter and setter match unless there’s a reason to do otherwise that I’m missing.  I think the reasoning that applied to correlationId also applies to headers, so I would lean towards using a Supplier in the setter.

### More General approach
Now you’ve got me thinking about how this could be generalized.  Both correlation id and api key are represented in HTTP headers.  Maybe we should maintain `Map<String, Supplier<String>>` that the `FormsFeederClient` uses to set header values.  It could loop through the entries in the Map and set a header for each one.

The Builder Interface would then have the following functions:
`Builder addHeader(String, Supplier<String>)`
`Map<String, Supplier<String>> getHeaderMap();`
The `FormsFeederClient` constructor would take an additional parameter -> `Map<String, Supplier<String>> headerMap`.  When it is building the invocation, it would loop through the headerMap and add the headers into the request.

Note: I've made the header name fixed and only have a `Supplier` for the value.  I am thinking that the headers you want to add will be known ahead of time and only the values will change from one invocation to the next.  Let me know if you think that is being too restrictive.

Let me know what you think of that idea.

### Bonus Points
While it’s not really necessary, you could re-implement the correlationId call to implemented as a default method on the Builder interface that calls `addHeader()` and remove both the getter from the Builder interface and the correlationId function parameter from the `FormsFeederClient` constructor.  This would simplify the code a bit while maintaining backwards compatibility for clients.

Reply:
> What immediately jumps out to me is the asymmetry here – using a name/value pair for the setter but getting a Supplier for the getter. Is there a reason for this?

Nope.. Truth be told, was just sort of backing into what I needed by starting somewhere and ended up with a strange mesh up of other things :)

> Let me know what you think of that idea.

I like the more general approach, I'll be submitting another commit shortly. I'm not entirely sure it's what you had in mind but I think it moves us closer.

>On a side note, my read of the PR #2 code has the Authentication key being passed in an HTTP header. Is this the same as the API key being passed as a query parameter above or are these two different parameters? If they are the same, does it need to be a query parameter or a http header? (I'm just looking to understand whether the need is to implement HTTP headers, query parameters or both.)

Quote from #3 , the apigateway actually allows either to be used, query param or header. I just decided to go the header route and was just using the query param for illustrative purposes. So the need is at least one of. But both may not be bad to have available.